LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan/TimeoutLinkedCancellationToken.cs

```diff
@@ -0,0 +1,63 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace LoRaWan
+{
+    using System;
+    using System.Threading;
+
+#pragma warning disable CA1001 // Types that own disposable fields should be disposable (false positive)
+    public readonly struct TimeoutLinkedCancellationToken : IEquatable<TimeoutLinkedCancellationToken>, IDisposable
+#pragma warning restore CA1001 // Types that own disposable fields should be disposable
+    {
+        private readonly CancellationTokenSource? timeoutCts;
+        private readonly CancellationTokenSource? linkedCts;
+
+        public CancellationToken Token { get; }
+
+        public TimeoutLinkedCancellationToken(TimeSpan? timeSpan, CancellationToken cancellationToken)
+        {
+            this.timeoutCts = this.linkedCts = null;
+
+            if (timeSpan is { } ts && ts != Timeout.InfiniteTimeSpan)
+            {
+                var tempTimeoutCts = this.timeoutCts = new CancellationTokenSource(ts);
+
+                try
+                {
+                    if (cancellationToken.CanBeCanceled)
+                    {
+                        this.linkedCts = CancellationTokenSource.CreateLinkedTokenSource(this.timeoutCts.Token, cancellationToken);
+                        Token = this.linkedCts.Token;
+                        tempTimeoutCts = null;
+                    }
+                    else
+                    {
+                        Token = this.timeoutCts.Token;
+                    }
+                }
+                finally
+                {
+                    tempTimeoutCts?.Dispose();
```

Comment:
I'm probably reading this wrong, but if it is a reference, we dispose the timeoutCts but on line 36 we could have assigned that token to the Token property. If I understand the implementation of the CancellationTokenSource correctly, the timeout would not fire when you dispose it. Hence the token wouldn't be canceled?

Reply:
The bug was fixed as part of a re-design of this.